feature/plot_recipes_for_IFF

mgyoo86 opened the pull request:
# Summary
This PR implements new `@recipes` for `IFF` (`IDS_Field_Finder`) structures.
The new recipes aim to plot multiple fields at once, i.e., multiple-subplots in a single figure, allowing users to quickly visualize and analyze the data of interest. 

Among the given IFF list, this recipe only plots either `AbstractVector{<:Real}` or `AbstractMatrix{<:Real}` data to avoid a possible conflict with other `@recipes` for specific IDS data.

- The new recipes have several optional arguments, which can be shown by `help_plot`.

# Feature description
There are two different fallbacks, depending on the datatype.

### 1.  `AbstractVector{<:Real}`
- This uses the existing recipe `@recipe function plot_field`
- Multiple fields call the `plot_field` recipe multiple times.

### 2. `AbstractMatrix{<:Real}`
- This calls one of the basic `seriestype` in the `Plots` package for 3-dimensional data `(x,y,z)`, such as `contourf`, `contour`, `heatmap`, and`surface`.
- Default `seriestype` is `contourf`, except for the `UnicodePlots` backend, where `contour` is set due to the compatibility issue.

# Optional arguments
- A recipe for `Vector{IMASdd.IDS_Field_Finder}`
  - `each_size`: Size of each subplot (Default = `(500,400)`)
  - `nrows`: Number of rows for subplots' layout (Default=`:auto`)
  - `ncols`: Number of columns for subplots' layout (Default=`:auto`)
  
- A recipe for `IMASdd.IDS_Field_Finder`
  - `abbreviations`: Abbreviations to shorten titles of subplots
  - `seriestype_2d`: Seriestype for 2D data (x,y)
  - `seriestype_3d`: Seriestype for 3D data (x,y,z)

See the following output with `help_plot` for example:

![Screenshot 2024-11-20 at 9 37 45 AM](https://github.com/user-attachments/assets/0a583d0e-e765-4ff1-9552-f427b23e6147)

# Examples
```julia
# Load IMASdd's sample json file
dd = IMAS.json2imas("IMASdd/sample/omas_sample.json"); 

# Find IFF and plot
IFF = findall(dd, [:psi, :j_tor]);
plot(IFF)

# There is a equivalent wrapper function for the user's convenience
plot(dd, [:psi,:j_tor]) # a wrapper for `findall` & `plot`
```

## 1. Default case
```julia
plot(dd, [:psi,:j_tor]) # a wrapper for `findall` & `plot`
```
![Screenshot 2024-11-19 at 5 36 54 PM](https://github.com/user-attachments/assets/69a6aad8-84bc-49e5-a69c-b1518002763c)

## 2. Control `each_size` and `ncols`
```julia
plot(dd, [:psi,:j_tor]; ncols=2, each_size=(300,300)) 
```
![Screenshot 2024-11-19 at 5 38 26 PM](https://github.com/user-attachments/assets/eb2b42af-759d-4368-8e5d-3819470c79bf)


## 3. `seriestype_3d` and `abbreviations`
 ```julia
# Use `Heatmap`for 3D data
# with no abbreviation for titles (by providing the empty `Dict`)
plot(dd, [:psi,:j_tor]; seriestype_3d=:heatmap, abbreviations=Dict()) 
```
![Screenshot 2024-11-19 at 5 39 20 PM](https://github.com/user-attachments/assets/55f78822-2745-43f0-9ba9-ac267fa77837)

## 4. More custom examples
```julia
# In addition to the `defulat_abbreviations`, 
# one can use the unicode or LaTexStrings for further polishing the titles
using LaTexStrings
new_abbrev = Dict(IMAS.default_abbreviations..., "psi"=>"ψ", "j_tor"=>L"j_{\phi}");
plot(dd, [:psi,:j_tor]; seriestype_2d=:bar, seriestype_3d=:surface, abbreviations=new_abbrev)
```

![Screenshot 2024-11-20 at 9 56 07 AM](https://github.com/user-attachments/assets/a4d84af0-e249-40b3-b4cb-3789e6f4f83e)

mgyoo86 commented:
I've updated some optional arguments, where I accidentally force-pushed the last commit.
Please let me know if it is better to re-open the pull request with a cleaner commit history.

orso82 commented on src/plot.jl:
```diff
@@ -2756,6 +2756,157 @@ end
+    return full_name
+end
+
+@recipe function plot_IFF(IFF::IDS_Field_Finder)
+    if Plots.backend_name() == :unicodeplots
+        seriestype_3d = get(plotattributes, :seriestype_3d, :contour)
+    else
+        seriestype_3d = get(plotattributes, :seriestype_3d, :contourf)
+    end
+
+    seriestype_2d = get(plotattributes, :seriestype_2d, :line)
+    abbreviations = get(plotattributes, :abbreviations, default_abbreviations)
```
```suggestion
@recipe function plot_IFF(IFF::IDS_Field_Finder; abbreviations=default_abbreviations, seriestype_2d=:line, seriestype_2d=:contourf)
    if Plots.backend_name() == :unicodeplots && seriestype_2d==:contourf
        seriestype_3d = get(plotattributes, :seriestype_3d, :contour)
    end
```

orso82 reviewed (changes requested):
👍 

Some minor comments

mgyoo86 commented:
## 1.  Added `@assert` in `assert_type_and_record_argument`

It seems like the current `assert_type_and_record_argument` does not check the type of argument. I added `@assert` which checks the type of argument.
Please review the commit [5b83d2a](https://github.com/ProjectTorreyPines/IMAS.jl/pull/216/commits/5b83d2ac67e36b575d0ed31a1fb9e9475ba61dcb)

## 2. The new optional argument `nicer_title`
`nicer_title` is a flag (Default=true) whether to further polish the title by using the combination of `nice_symbol` and `nice_units` as you suggested.
See the following examples:
### 2.1. `nicer_title=true` (Default)
![Screenshot 2024-11-21 at 11 13 45 AM](https://github.com/user-attachments/assets/7ff5c5db-169c-43f6-80e3-aa6c25e6953d)

### 2.2. `nicer_title=false`
![Screenshot 2024-11-21 at 11 13 54 AM](https://github.com/user-attachments/assets/4548634e-2d96-4e4e-8ca2-58dba60f6c1d)


## 3. Behavior change for `Matrix` data (3D plot)
For `AbstractMatrix{<:Real}` data, it first checks if its `parent_ids` has valid `grid` dimensions, i.e. `grid.dim1` and `grid.dim2` has sorted ascending values.
If the `parent_ids` have valid grid dimensions, it uses them with `aspect_ratio=:equal`.
Otherwise, the recipe shows `Matrix` data, denoted by `row` and `column`, without `aspect_ratio`.

For example, regarding two 3D plots in the above screenshots, 
the left-one (`eq.time_sc[2].prof_2d[1].psi [Wb]`), (which are MXH coeffecients) shows the  `matrix` data, and the right-one (`eq.time_sc[2].prof_2d[2].psi [Wb]`) shows the psi in the (R,Z) geometry.

orso82 commented on src/plot.jl:
```diff
@@ -35,6 +35,8 @@ function assert_type_and_record_argument(dispatch, type::Type, description::Stri
     argument = collect(keys(kw))[1]
     value = collect(values(kw))[1]
 
+    @assert typeof(value) <: type "\"$argument\" has the wrong type of [$(typeof(value))]. It must be a subtype of [$(type)]"
```
Ah, I thought we did check for the type. Thank you for adding this @mgyoo86

orso82 commented:
@mgyoo86 don't hardcode looking for `grid`. Instead you can use `coordinates(ids, field)` to get the coordinate of a field.

mgyoo86 commented:
@orso82 Since "coord.names" is too long for labels, I used the last field (use `split` with the delimiter of ".") for xlabel and ylabel.

orso82 commented:
@mgyoo86 could we please use `Measures.mm` instead of `using Plots.PlotMeasures`. Note that IMAS already does a `import Measures`.

In general we try to avoid `using` since that pollutes the local namespace and makes it hard to know where a function is defined.

PS. Originally IMAS was designed to not have `Plots.jl` as a dependency. However, this made debugging difficult, since plotting was not available. It's debatable whether we should remove the individual packages `PlotUtils`, `RecipesBase`, `Measures` or just use `Plots.jl`.

mgyoo86 commented:
@orso82 Sorry, I didn't know that Measures is already implemented. I fixed it as you recommended!

orso82 commented:
@mgyoo86 thank you for bearing with me on this one :)

I'll change the base branch for this PR from `master` to the `documentation` branch, so that we can see if there's any conflict that we need to deal with (we might be ok, I think). I want to merge the `documentation` branch by the end of today.

orso82 commented:
No conflicts 🎉  I'll merge it now into the `documentation` branch.
